### PR TITLE
gh-115554: Improved logic for handling multiple existing py.exe launcher installs

### PR DESCRIPTION
**Pull request by zooba**
* Issue: gh-115554

### CODE REVIEW

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp [hunk 1]
```diff
@@ -718,25 +726,64 @@ class PythonBootstrapperApplication : public CBalBaseBootstrapperApplication {
...
+        // Only check launcher_AllUsers because we'll find the same packages
+        // twice if we check launcher_JustForMe as well.
+        if (CSTR_EQUAL == ::CompareStringW(LOCALE_NEUTRAL, 0, wzPackageId, -1, L"launcher_AllUsers", -1)) {
+            BalLog(BOOTSTRAPPER_LOG_LEVEL_STANDARD, "Detected existing launcher install");
+
+            LONGLONG blockedLauncher, detectedLauncher;
+            if (FAILED(BalGetNumericVariable(L"BlockedLauncher", &blockedLauncher))) {
+                blockedLauncher = 0;
+            }
+            if (FAILED(BalGetNumericVariable(L"DetectedLauncher", &detectedLauncher))) {
```
Comment (gpshead) [inline Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:742]: Very windows installer API code naively I'd think looking up a DetectedLauncher numeric variable, then setting it to 1 if it is not already true is equivalent to just always setting it to 1?

add a comment as to why this pattern is necessary otherwise.

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp [hunk 2]
```diff
@@ -718,25 +726,64 @@ class PythonBootstrapperApplication : public CBalBaseBootstrapperApplication {
...
+            if (blockedLauncher) {
+                // Nothing else to do, we're already blocking
+            }
+            else if (BOOTSTRAPPER_RELATED_OPERATION_DOWNGRADE == operation) {
+                // Found a higher version, so we can't install ours.
+                BalLog(BOOTSTRAPPER_LOG_LEVEL_ERROR, "Higher version launcher has been detected.");
+                BalLog(BOOTSTRAPPER_LOG_LEVEL_ERROR, "Launcher will not be installed");
+                _engine->SetVariableNumeric(L"BlockedLauncher", 1);
+            }
+            else if (detectedLauncher == 1) {
```
Comment (gpshead) [inline Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp]: !detectedLauncher was used above for == 0, this uses == 1, should it just use `if (detectedLauncher)` instead or are there other values than 0/1 to be concerned with?

File: Misc/NEWS.d/next/Windows/2024-02-21-23-48-59.gh-issue-115554.02mpQC.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+The installer now has more strict rules about updating :ref:`launcher`. In
```
Comment (itamaro) [inline Misc/NEWS.d/next/Windows/2024-02-21-23-48-59.gh-issue-115554.02mpQC.rst]:
```suggestion
The installer now has more strict rules about updating the :ref:`launcher`. In
```

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp [hunk 4]
```diff
@@ -718,25 +726,64 @@ class PythonBootstrapperApplication : public CBalBaseBootstrapperApplication {
...
+                }
             }
+            else if (BOOTSTRAPPER_RELATED_OPERATION_MAJOR_UPGRADE == operation) {
+                // Found an older version, so let's run the equivalent as an upgrade
+                // This overrides "unknown" all users options, but will leave alone
+                // any that have already been set/detected.
+                // User can deselect the option to include the launcher, but cannot
+                // change it from the current per user/machine setting.
+                LONGLONG includeLauncher, includeLauncherAllUsers;
+                if (FAILED(BalGetNumericVariable(L"Include_launcher", &includeLauncher))) {
```
Comment (gpshead) [inline Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:772]: The _ in that L" Variable name when most others don't appear use that style it is unfortunate... but i'm guessing this comes from elsewhere in the XML definition or whatnot and is just style would require larger refactoring beyond the scope of this PR better done as its own no-functionality change if ever.

**Review by gpshead — COMMENTED** [review]
at a high level I think this makes sense.  I can look past the specific api c++ boilerplate for what really seems to just be a way to spell get and set these variables from the whatever-script maybe functionalish installer state machine runtime ultimately running this logic.

My comments are mostly code stuff, I trust that this does what it describes itself as doing.  But I'm definitely not a windows installer expert (I last created one over 20 years ago).

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp [hunk 5]
```diff
@@ -718,25 +726,64 @@ class PythonBootstrapperApplication : public CBalBaseBootstrapperApplication {
...
-            if (hr == S_OK) {
-                _engine->SetVariableNumeric(L"AssociateFiles", 1);
-            } else if (hr == S_FALSE) {
-                _engine->SetVariableNumeric(L"AssociateFiles", 0);
-            } else if (FAILED(hr)) {
-                BalLog(BOOTSTRAPPER_LOG_LEVEL_ERROR, "Failed to load AssociateFiles state: error code 0x%08X", hr);
+        // Only check launcher_AllUsers because we'll find the same packages
+        // twice if we check launcher_JustForMe as well.
+        if (CSTR_EQUAL == ::CompareStringW(LOCALE_NEUTRAL, 0, wzPackageId, -1, L"launcher_AllUsers", -1)) {
+            BalLog(BOOTSTRAPPER_LOG_LEVEL_STANDARD, "Detected existing launcher install");
```
Comment (itamaro) [inline Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:732]: could the log include more info about the existing installation? (like version, location)

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp [hunk 6]
```diff
@@ -718,25 +726,64 @@ class PythonBootstrapperApplication : public CBalBaseBootstrapperApplication {
...
+            if (!detectedLauncher) {
+                _engine->SetVariableNumeric(L"DetectedLauncher", 1);
+            }
+
+            if (blockedLauncher) {
+                // Nothing else to do, we're already blocking
+            }
+            else if (BOOTSTRAPPER_RELATED_OPERATION_DOWNGRADE == operation) {
+                // Found a higher version, so we can't install ours.
+                BalLog(BOOTSTRAPPER_LOG_LEVEL_ERROR, "Higher version launcher has been detected.");
```
Comment (itamaro) [inline Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:754]: can this include more info about the version that was detected?

**Review by itamaro — COMMENTED** [review]
this seems a bit convoluted and hard to follow (maybe a combination of installer-toolkit-isms and non-trivial business logic) -- it might benefit from a flowchart!

can you clarify what would be the behavior when both 3.12.1 & 3.11.5 are installed, and we attempt to upgrade both to 3.12.2 & 3.11.8? (arbitrary patch releases)
in particular, I'm curious about the scenario of unattended installations that attempt to upgrade both (and can happen in unspecified order on different machines).

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp (hunk 5 above)
Comment (zooba) [reply Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:732]: That's included in a log message that appears immediately before this function is called, so no need.

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp (hunk 6 above)
Comment (zooba) [reply Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:754]: Included in an automatic log message already (you can see examples in the linked issue for this PR)

**Comment by zooba** [comment]
> this seems a bit convoluted and hard to follow

Agreed! It hurt my head figuring out the right places to put the logic.

> can you clarify what would be the behavior when both 3.12.1 & 3.11.5 are installed, and we attempt to upgrade both to 3.12.2 & 3.11.8? (arbitrary patch releases)

So the important aspect for this change isn't which Python version(s) are installed, but which _launcher_ version. There's only one launcher installer, so there should only ever be one version installed. Multiple is a bad enough error that we now refuse to do anything.

If we assume 3.12.1 and 3.11.5 are installed correctly, then launcher 3.12.1 will be installed. If we then install 3.11.8, it'll compare its launcher 3.11.8 to 3.12.1 and decide that's a _downgrade_, and so will take the downgrade path in the new code (which deselects the launcher and disables the controls).

If we're installing 3.12.2, it'll compare its launcher 3.12.2 to 3.12.1 and decide it's a _major upgrade_, and so will select the launcher and update the "launcher for all users" option to match the existing install (but only if the option wasn't specified on the command line, in which case we disable the launcher entirely if they conflict).

If a second launcher is installed, this typically means one is per-user and one is per-machine, which we can't handle, so we block the launcher install entirely. This will update the checkbox on the main page to say that the launcher has to be uninstalled in order to update.

If no launchers are detected, we update the default options (-1) to what we want for our standard default (i.e. to include the launcher). It's possible to override these on the command line though, so we want to respect those selections.

It's not much of a flowchart, but you can probably imagine "OnPackageDetected" getting called once for every package that's already installed, and "OnDetectionComplete" being called once at the end.

File: Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp (hunk 4 above)
Comment (zooba) [reply Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:772]: It's actually consistent with the other `Include_<name>` variables, which all refer to a `<name>.msi` file. It's just none of the others need to be referenced as much because they're all easy.

Comment (zooba) [reply Tools/msi/bundle/bootstrap/PythonBootstrapperApplication.cpp:772]: Here they are all in a list: https://github.com/python/cpython/blob/41d5391c551e2012f13d56ff4bcc1df15c2821d3/Tools/msi/bundle/bundle.wxs#L68-L77

**Review by itamaro — APPROVED** [review]
the Windows installer could probably benefit from an end-to-end test suite, with all the complexity and possible scenarios, but I can't imagine writing such a test suite would be easy, and it'll probably be excruciatingly slow and flaky...

**Comment by zooba** [comment]
> I can't imagine writing such a test suite would be easy, and it'll probably be excruciatingly slow and flaky...

I just tried getting the smoke tests that are part of the installer build to run more reliably (GitHub installs Python on the build machines, so the installers can't really be tested when they're already installed 😅) and yeah, slow and flaky. But we do run the basics through their paces on every release - just not through every possible combination.

However, most combinations shouldn't have interactions, and the problems almost always come up because of preexisting things on peoples' machines, and not because of the installer itself. So it becomes a case of scripting the setup of broken machines so that we can run the test and ... err... yeah... it's a huge amount of work for not actually a lot of benefit.
